flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBKeyedStateBackendBuilder.java

```diff
@@ -0,0 +1,502 @@
+				IOUtils.closeQuietly(cancelStreamRegistry);
+			}
+			try {
+				FileUtils.deleteDirectory(instanceBasePath);
+			} catch (Exception ex) {
+				LOG.warn("Failed to instance base path for RocksDB: " + instanceBasePath, ex);
+			}
+			// Log and rethrow
+			if (e instanceof BackendBuildingException) {
+				throw (BackendBuildingException) e;
+			} else {
+				String errMsg = "Caught unexpected exception.";
+				LOG.error(errMsg, e);
+				throw new BackendBuildingException(errMsg, e);
+			}
+		}
+	}
+
+	private void initializeSnapshotStrategy(CloseableRegistry cancelStreamRegistry,
+											ResourceGuard rocksDBResourceGuard,
+											LinkedHashMap<String, StateColumnFamilyHandle> kvStateInformation,
+											int keyGroupPrefixBytes) {
+		this.savepointSnapshotStrategy = new RocksFullSnapshotStrategy<>(
+			db,
+			rocksDBResourceGuard,
+			keySerializer,
+			kvStateInformation,
+			keyGroupRange,
+			keyGroupPrefixBytes,
+			localRecoveryConfig,
+			cancelStreamRegistry,
+			keyGroupCompressionDecorator);
+		if (enableIncrementalCheckpointing) {
+			// TODO eventually we might want to separate savepoint and snapshot strategy, i.e. having 2 strategies.
+			checkpointSnapshotStrategy = new RocksIncrementalSnapshotStrategy<>(
+				db,
+				rocksDBResourceGuard,
+				keySerializer,
+				kvStateInformation,
+				keyGroupRange,
+				keyGroupPrefixBytes,
+				localRecoveryConfig,
+				cancelStreamRegistry,
+				instanceBasePath,
+				backendUID,
+				materializedSstFiles,
+				lastCompletedCheckpointId,
+				numberOfTransferingThreads);
+		} else {
+			checkpointSnapshotStrategy = savepointSnapshotStrategy;
+		}
+	}
+
+	private PriorityQueueSetFactory initPriorityQueueFactory(RocksDBKeyedStateBackend<K> backend) {
+		PriorityQueueSetFactory priorityQueueFactory;
+		switch (priorityQueueStateType) {
+			case HEAP:
+				priorityQueueFactory = new HeapPriorityQueueSetFactory(keyGroupRange, numberOfKeyGroups, 128);
+				break;
+			case ROCKSDB:
+				priorityQueueFactory = new RocksDBPriorityQueueSetFactory(backend);
+				break;
+			default:
+				throw new IllegalArgumentException("Unknown priority queue state type: " + priorityQueueStateType);
+		}
+		return priorityQueueFactory;
+	}
+
+	private void postBackendInitialization(
+		RocksDBKeyedStateBackend<K> backend,
+		RocksDBIncrementalRestorePrepareResult restorePrepareResult) throws BackendBuildingException {
+		// Do necessary db clip and CF handlers registration
+		KeyGroupRange initialHandleKeyGroupRange = restorePrepareResult.getInitialHandleKeyGroupRange();
+		if (initialHandleKeyGroupRange != null) {
+			// 1. Restore with rescaling, do clip if the DB instance is restored from an initial handle
+			try {
+				RocksDBIncrementalCheckpointUtils.clipDBWithKeyGroupRange(
+					db,
+					columnFamilyHandles,
+					keyGroupRange,
+					initialHandleKeyGroupRange,
+					backend.getKeyGroupPrefixBytes());
+			} catch (RocksDBException e) {
+				backend.dispose();
```

Comment:
Looks like this is not required and we call dispose twice, the second time is in line 276.